adagrams/game.py

```diff
@@ -1,11 +1,175 @@
+#   POOL = gen_pool_letters(POOL_RULES)
+
 def draw_letters():
-    pass
+    draw = list()
+    my_ran_pool = gen_pool_letters(POOL_RULES)
+    
+    #generate random
+    random.seed()
+    for i in range(0,SIZE_OF_DRAW):
+        ran_num = random.randint(0,len(my_ran_pool)-1)
+        draw.append(my_ran_pool[ran_num])
+        my_ran_pool.pop(ran_num)
+
+    return draw
+
+#creates a dictionary which keys are the elements on the letters array 
+#and the value is their frecuency
+def create_dic_repeatead_letters(letters):
+    letters_dict = dict()
+
+    for letter in letters:
+        if letter in letters_dict.keys():
+            letters_dict[letter] += 1
+        else:
+            letters_dict[letter] = 1
+
+    return letters_dict
 
 def uses_available_letters(word, letter_bank):
-    pass
+    #Validate input
+    if word == "" or letter_bank is None:
+        return False
+
+    #Create a dictionary with the number of each letter of the letter bank
+    letter_bank_dict = create_dic_repeatead_letters(letter_bank)
+
+    #Compare each letter of word with the dictionary and their frecuency
+    for letter in word:
+        up_letter = letter.upper()
+        if up_letter not in letter_bank_dict.keys():
+            return False
+        if letter_bank_dict[up_letter]==0:
+            return False  #The letter has been used more times than allowed
+        else:
+            letter_bank_dict[up_letter]+=-1
+
+    return True
 
 def score_word(word):
-    pass
+    points = 0
+    if word is None:
+        return points
+
+    for letter in word:
+        letter_up = letter.upper()
+        if letter_up not in LETTERS_VALUES.keys():
+            return 0
+        points += LETTERS_VALUES[letter_up]
+
+    word_len = len(word)
+    if word_len>=7 and word_len<=10 :
+        points+=8
+
+    return points
 
 def get_highest_word_score(word_list):
-    pass
\ No newline at end of file
+    if word_list is None: 
+        return "",0
+
+    scores = dict()
+    max_score = 0
+    winner_word = ""
+    for word in word_list:
+        score = score_word(word)
+        if score>max_score:
+            max_score = score
+            winner_word = word
+            scores[score] = list()
+            scores[score].append(word)
+        elif score == max_score:
+            scores[score].append(word)
+
+    #There are words with the same score
+    if len(scores[max_score])> 1:
+        min_len = WINNER_LEN-1
+        words_same_len = dict()
+        for word in scores[max_score]:
+            lenght = len(word)
+            if lenght == WINNER_LEN:
+                return word, max_score
+            if lenght < min_len:
+                min_len = lenght
+                winner_word = word
+                words_same_len[min_len] = list()
+                words_same_len[min_len].append(word)
+                continue
+            if lenght == min_len: #there are more words with the same lenght
+                words_same_len[min_len].append(word)
```

Comment:
Since we want to return the first word that we find with a particular length when we're looking for the shortest word, if we needed to optimize for memory, we could keep a variable that holds a single word over a list like `words_same_len` that holds all the words with that length.